src/components/Conversation.vue

```diff
@@ -9,32 +9,177 @@
         </div>
       </div>
     </div>
-    <div id="conversation-body" class="body" contenteditable="true">
+    <div
+      id="conversation-body"
+      class="body"
+      contenteditable="true"
+      v-on:input="sendPatch"
+    >
     </div>
   </div>
 </template>
 
 <script>
+import DiffMatchPatch from 'diff-match-patch';
+
+// Declare constants
+const initialProps = Object.freeze(['content', 'version']);
+const updateProps = Object.freeze(['type', 'version', 'patch', 'cursor_delta']);
+const updateTypes = Object.freeze({
+  Edit: 0,
+});
+
+const dmp = new DiffMatchPatch();
+
 const data = () => ({
-  conversation: null,
+  ws: null,
+  conversation: {},
+  checkpoint: '',
+  version: 0,
+  content: '',
+  patchBuffer: [],
 });
 
+const checkMsgProps = (msg, requiredProps) => {
+  const missingProps = [];
+  requiredProps.forEach((prop) => {
+    if (!(prop in msg)) {
+      missingProps.push(prop);
+    }
+  });
+
+  return missingProps.join();
+};
+
 function conversations() {
   return this.$store.state.conversations;
 }
 
 function activeConversation() {
-  return this.$store.state.activeConversation;
+  return parseInt(this.$route.params.id, 10);
+}
+
+function sendPatch() {
+  const patches = dmp.patch_make(this.content, this.conversationDOM.innerHTML);
+  patches.forEach((patch) => {
+    this.version += 1;
+    const patchStr = dmp.patch_toText([patch]);
+    if (patchStr === '') {
+      return;
+    }
+
+    const msg = {
+      type: updateTypes.Edit,
+      version: this.version,
+      patch: patchStr,
+      cursor_delta: 0,
+    };
+
+    this.patchBuffer.push(patchStr);
+    this.ws.send(JSON.stringify(msg));
+  });
+
+  this.content = this.conversationDOM.innerHTML;
+}
+
+function parseWSMessage(e) {
+  let msg = {};
+  try {
+    msg = JSON.parse(e.data);
+  } catch (err) {
+    console.log(`Unable to parse JSON data: ${err.message}`);
+    return;
+  }
+
+  if ('content' in msg) { // Initial message
```

Comment:
Instead of relying on a field being in the message to determine what to do, it might be better to either have a `waitingForInitMessage` boolean or just set the `version` to -1 initially and just check for that instead.

Reply:
Yeah I just wasn't sure if we're planning on having more types of messages but I guess it'd just be the initial message and updates right?